base/libgit2/callbacks.jl

```diff
@@ -31,74 +31,131 @@ If payload is provided then `payload_ptr` should contain `LibGit2.AbstractPayloa
 
 For `LibGit2.Consts.CREDTYPE_USERPASS_PLAINTEXT` type, if payload contains fields: `user` & `pass` they are used to create authentication credentials.
 In addition, if payload has field `used` it can be set to `true` to indicate that payload was used and abort callback with error. This behavior is required to avoid repeated authentication calls with incorrect credentials.
+
+Order of credential checks (if supported):
+- ssh key pair (ssh-agent if specified in `payload`)
+- plain text
+
+**Note**: Due to the specifics of `libgit2` authentication procedure, when authentication is failed, this functions is called again without any indication whether authentication was successful or not. In order not to stuck in infinite loop by repeatedly using same faulty credentials, use a call counting strategy to counteract this behaviour.
 """
 function credentials_callback(cred::Ptr{Ptr{Void}}, url_ptr::Cstring,
                               username_ptr::Cstring,
                               allowed_types::Cuint, payload_ptr::Ptr{Void})
-    err = 1
+    err = 0
     url = bytestring(url_ptr)
 
-    if isset(allowed_types, Cuint(Consts.CREDTYPE_USERPASS_PLAINTEXT))
-        username = userpass = ""
-        if payload_ptr != C_NULL
-            payload = unsafe_pointer_to_objref(payload_ptr)
-            if isa(payload, AbstractPayload)
-                isused(payload) && return Cint(-1)
-                username = user(payload)
-                userpass = password(payload)
-                setused!(payload, true)
-            end
+    # parse url for schema and host
+    urlparts = match(urlmatcher, url)
+    schema = urlparts.captures[1]
+    host = urlparts.captures[5]
+    schema = schema === nothing ? "" : schema*"://"
+
+    # get credentials object form payload ref
+    creds = EmptyCredentials()
+    if payload_ptr != C_NULL
+        tmpobj = unsafe_pointer_to_objref(payload_ptr)
+        if isa(tmpobj, AbstractCredentials)
+            creds = tmpobj
         end
-        if isempty(username)
-            username = prompt("Username for '$url'")
-        end
-        if isempty(userpass)
-            userpass = prompt("Password for '$username@$url'", password=true)
+    end
+    isusedcreds = isused(creds)
+
+    # use ssh key or ssh-agent
+    if isset(allowed_types, Cuint(Consts.CREDTYPE_SSH_KEY))
+        credid = "ssh://$host"
+        # first try ssh-agent if credentials support its usage
+        if isdefined(creds, :use_ssh_agent) && getfield(creds, :use_ssh_agent) == 0x00
```

Comment:
can't the `getfield` and `setfield!` just using `creds.use_ssh_agent` if it's guarded by an `isdefined`?